torch/fx/experimental/Partitioner.py

```diff
@@ -106,77 +122,378 @@ def get_output_nodes(self) -> List[Node]:
     so that each partition could fit into a different device.
+    The main function of this class is self.partition_graph.
+    For self.partition_graph, first, it checks the size of the whole graph
+    and see if the whole graph can fit into one device.
+    If it does, it goes to self.find_single_partition
+    If the whole graph is even larger than the combined memory size of all devices,
+    a RuntimeError is raised.
+    If the whole graph cannot fit into one devices but
+    could be split into multiple devices, it goes to self.size_based_partition.
+    After the size_based_partition, it checks if the number of partitions exceeds
+    the number of devices. If it does, a RuntimeError is raised.
+    Otherwise, a DAG structure is returned
+    along with a new graph module with partitions as submodule nodes.
     """
     def __init__(self) -> None:
-        """
-        After a partitioner is created,
-        it first check if multiple types of devices (backends) are involved.
-        So far, we assume that there is only one backend and one device with unlimited memory
-        """
-        self.partitions: List[Partition] = []
-        self.devices: List[Dict[str, Union[str, float]]] = []
+        self.partitions: Set[Partition] = set()
+        self.devices: List[Device] = []
         self.node_to_partitions: Dict[Node, List[int]] = {}
+        self.partition_to_used_mem: Dict[Partition, int] = {}
+        self.orphans: Set[Node] = set()
+        self.args_mapping: Dict[Node, List[Node]] = {}
+        self.users_mapping: Dict[Node, List[Node]] = {}
 
-    def partition_graph(self, fx_module: GraphModule, devices: List[Dict[str, Union[str, float]]]) -> DAG:
+    def partition_graph(
+        self,
+        fx_module: GraphModule,
+        torch_module: torch.nn.Module,
+        devices: List[Device]
+    ) -> PartitionResult:
         """
-        Given the fx_module and devices, find the partitions, do the partitions,
-        and then return a dictionary representing the DAG of partitions
+        Given the fx module, torch module and devices,
+        find the partitions, do the partitions,
+        and then return a DAG and a new fx module with submodule nodes (partitions)
         """
         self.graph_module = fx_module
         self.devices = devices
-        # Create a dummy partition as root.
-        self.root_partition = self.create_partition()
-        # So far, the whole could fit into one device since we assume the memory is unlimited.
-        # TODO: Check the available memory size for each device and see if one device is enough or multiple partitions are needed.
-        self.find_single_partition()
-        self.do_partition()
-        dag = self.dump_partition_DAG()
-        return dag
+        self.torch_module = torch_module
+        if len(self.devices) == 0:
+            raise RuntimeError('No devices')
+        available_mem = self.devices[0].available_mem_bytes
```

Comment:
nit: make this variable name `available_mem_bytes` as well, and maybe `total_size_of_graph_bytes` below